### PR TITLE
[FEATURE] V3 - Implement multicolumn helpers/providers and MulticolumnSumEqual metric for SQL Alchemy

### PR DESCRIPTION
**Pull request by NathanFarmer**
Changes proposed in this pull request:
- JIRA GREAT-56/GREAT-4
- JIRA GREAT-56/GREAT-11
- JIRA GREAT-56/GREAT-12

### Definition of Done

- [X] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [X] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [X] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

**Review by alexsherstinsky — CHANGES_REQUESTED** [review]
@NathanFarmer This looks logically correct to me.  Here are some observations:
A. I suspect that the reason the errors are happening is because the key metrics do not have an SQL implementation, while the expectations are not suppressed for the SQL backend (please see a little bit more on that below).

B. Next, I would propose the following steps in order get this PR to a mergeable state:
1) Wait till the "IDENTITY" PRs get merged so that you can merge the result into your PR and resolve conflicts (happy to pair on this exercise if you think that it will make things go faster).
2) Update the `MulticolumnSumEqual` metric to include the usage of your new SQL condition partial.
3) Add a unit test for the SQL version of this metric to `tests/expectations/metrics/test_core.py` as the sibling of the corresponding `Pandas` test.

Then, in the next PR, the following:
4) Enable the integration test of the `expect_multicolumn_sum_to_equal` expectation on `sqlalchemy` and all databases and verify that it works locally (again, happy to collaborate on this, if that would be helpful to you).

Please let me know what you think.

Thank you!

**Comment by NathanFarmer** [comment]
@alexsherstinsky

> 2. Update the MulticolumnSumEqual metric to include the usage of your new SQL condition partial.

Is this referring to the changes in `multicolumn_sum_equal.py`? I split these out after your previous feedback regarding splitting up the PR. I am happy to merge the changes back into this branch if it is needed. I just want to be precise about what exactly this means.

**Comment by NathanFarmer** [comment]
@alexsherstinsky The tests are now passing without merging in the SQL Alchemy implementation in `multicolumn_sum_equal.py` or the unit test added I to `test_core.py`. I would like to propose that the helpers/providers are merged into develop as-is (after I pull in your IDENTITY changes in PR [#3215](https://github.com/great-expectations/great_expectations/pull/3215)). I will then open a second PR for steps B. 2-4 you outlined above.

**Comment by NathanFarmer** [comment]
I will be merging in all changes listed below using this PR following an offline conversation with @alexsherstinsky.

> 1. Wait till the "IDENTITY" PRs get merged so that you can merge the result into your PR and resolve conflicts (happy to pair on this exercise if you think that it will make things go faster).
> 2. Update the `MulticolumnSumEqual` metric to include the usage of your new SQL condition partial.
> 3. Add a unit test for the SQL version of this metric to `tests/expectations/metrics/test_core.py` as the sibling of the corresponding `Pandas` test.
> 4. Enable the integration test of the `expect_multicolumn_sum_to_equal` expectation on `sqlalchemy` and all databases and verify that it works locally (again, happy to collaborate on this, if that would be helpful to you).

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 1]
```diff
@@ -1085,6 +1208,32 @@ def _pandas_map_condition_rows(
...
+    metrics: Dict[str, Any],
+    **kwargs,
+):
+    """Returns unexpected count for MapExpectations"""
+    (
+        boolean_mapped_unexpected_values,
+        compute_domain_kwargs,
+        accessor_domain_kwargs,
+    ) = metrics["unexpected_condition"]
+
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/map_metric_provider.py]: @NathanFarmer Super minor: Let us delete line 1225 (blank line) to match the style of the rest.  Thanks!

File: great_expectations/expectations/metrics/map_metric_provider.py [hunk 2]
```diff
@@ -1669,6 +1900,34 @@ def _register_metric_functions(cls):
...
                             metric_fn_type=MetricFunctionTypes.VALUE,
                         )
+                    elif domain_type == MetricDomainTypes.MULTICOLUMN:
+                        register_metric(
+                            metric_name=metric_name + ".unexpected_count",
+                            metric_domain_keys=metric_domain_keys,
+                            metric_value_keys=metric_value_keys,
+                            execution_engine=engine,
+                            metric_class=cls,
+                            metric_provider=_sqlalchemy_map_condition_unexpected_count,
```
Comment (alexsherstinsky) [inline great_expectations/expectations/metrics/map_metric_provider.py]: @NathanFarmer Do we need this?  I notice that we already register the `unexpected_count` metric earlier.  If you agree, then we can delete the `_sqlalchemy_map_condition_unexpected_count()` helper method.  Happy to discuss.

File: tests/expectations/metrics/test_core.py [hunk 3]
```diff
@@ -1769,7 +1908,7 @@ def test_map_compound_columns_unique():
     )
     metrics.update(results)
 
-    # Condition metrics return "negative logic" series.
+    # Condition metrics return "negative logic" series
```
Comment (alexsherstinsky) [inline tests/expectations/metrics/test_core.py]: @NathanFarmer Minor: The comment in the `Pandas` test has the period at the end, so let us keep it here as well.  Thanks.

File: tests/test_definitions/multicolumn_map_expectations/expect_multicolumn_sum_to_equal.json [hunk 4]
```diff
@@ -35,7 +35,7 @@
       }
     },{
       "title": "row_condition_with_ignore_if_any_are_missing",
-      "only_for": ["pandas"],
+      "suppress_test_for": "spark",
```
Comment (alexsherstinsky) [inline tests/test_definitions/multicolumn_map_expectations/expect_multicolumn_sum_to_equal.json:38]: @NathanFarmer Please confirm what I am seeing: your PR enables the metric for `SQL`, hence the expectation is automatically enabled on `SQL`, thereby allowing us to make the statement that only the `Spark` integration test needs to be suppressed.  Thanks.

File: great_expectations/expectations/metrics/map_metric_provider.py (hunk 2 above)
Comment (NathanFarmer) [reply great_expectations/expectations/metrics/map_metric_provider.py]: Without helper `_sqlalchemy_map_condition_unexpected_count()` and registering this `unexpected_count` in this block I get: 
```
great_expectations.exceptions.exceptions.MetricProviderError: No provider found for multicolumn_sum.equal.unexpected_count using SqlAlchemyExecutionEngine
```

File: tests/test_definitions/multicolumn_map_expectations/expect_multicolumn_sum_to_equal.json (hunk 4 above)
Comment (NathanFarmer) [reply tests/test_definitions/multicolumn_map_expectations/expect_multicolumn_sum_to_equal.json:38]: @alexsherstinsky You are correct, and I have verified that logic in `test_expectations_cfe.py` and `test_expectations.py`.

File: great_expectations/expectations/metrics/map_metric_provider.py (hunk 2 above)
Comment (NathanFarmer) [reply great_expectations/expectations/metrics/map_metric_provider.py]: If I remove the registration for the `unexpected_count` that starts at line 1862 and has `metric_provider=None`, `pytest` starts failing with various `KeyError`s for `result["unexpected_list"]`.

Comment (NathanFarmer) [reply great_expectations/expectations/metrics/map_metric_provider.py]: @alexsherstinsky Following our offline conversation I have removed the duplicate helper and changed the `metric_provider` to `_sqlalchemy_map_condition_unexpected_count_value`.

**Review by alexsherstinsky — APPROVED** [review]
@NathanFarmer This looks good to me.  It would be good to also have a glance-over by @jcampbell for your first major PR!  Also, if we could please wait for the PR #3226 to be reviewed and merged first (within reasonable time), that would make things easier down the road.  Thank you!

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you very much @NathanFarmer for this PR 🔥  I really appreciated how you laid out the associated unittest along with inline comments. LGTM to me as well

**Pull request merged**